src/mlpack/core/optimizers/fw/update_span.hpp

```diff
@@ -0,0 +1,132 @@
+
+#include <mlpack/prereqs.hpp>
+#include "func_sq.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+/**
+ * Recalculate the optimal solution in the span of all previous solution space,
+ * used as update step for FrankWolfe algorithm.
+ *
+ * Currently only works for function in FuncSq class.
+ *
+ */
+class UpdateSpan
+{
+ public:
+  /**
+   * Construct the span update rule. The function to be optimized is 
+   * input here.
+   *
+   * @param function Function to be optimized in FrankWolfe algorithm.
+   */
+  UpdateSpan()
+  { /* Do nothing. */ }
+
+  /**
+   * Update rule for FrankWolfe, reoptimize in the span of current
+   * solution space.
+   *
+   * @param function function to be optimized.
+   * @param oldCoords previous solution coords, not used in this update rule.
+   * @param s current linearConstrSolution result.
+   * @param newCoords output new solution coords.
+   * @param numIter current iteration number.
+   */
+  void Update(FuncSq& function,
+              const arma::mat& oldCoords,
+              const arma::mat& s,
+              arma::mat& newCoords,
+              const size_t numIter)
+  {
+    // Add new atom into soluton space.
+    arma::uvec ind = find(s, 1);
+    arma::uword d = ind(0);
+    AddAtom(function, d);
+
+    // Reoptimize the solution in the current space.
+    arma::vec b = function.Vectorb();
+    arma::mat x = solve(currentAtoms, b);
+
+    // x has coords of only the current atoms, recover the solution
+    // to the original size.
+    newCoords = RecoverVector((function.MatrixA()).n_cols, x);
+  }
+
+  //! Get the current atom indices.
+  const arma::uvec& CurrentIndices() const { return currentIndices; }
+  //! Modify the current atom indices.
+  arma::uvec& CurrentIndices() { return currentIndices; }
+
+  //! Get the current atoms.
+  const arma::mat& CurrentAtoms() const { return currentAtoms; }
+  //! Modify the current atoms.
+  arma::mat& CurrentAtoms() { return currentAtoms; }
+
+ private:
+  //! Indices of the atoms in the current solution space.
+  arma::uvec currentIndices;
+
+  //! Current atoms in the solution space, ordered as currentIndices.
+  arma::mat currentAtoms;
+
+  //! Flag current indices is empty.
+  bool isEmpty = true;
+
+  //! Add atom into the solution space.
+  void AddAtom(FuncSq& function, const arma::uword k)
+  {
+    if (isEmpty)
+    {
+      CurrentIndices() = k;
+      CurrentAtoms() = (function.MatrixA()).col(k);
+      isEmpty = false;
+    }
+    else
+    {
+      arma::uvec vk(1);
+      vk = k;
+      currentIndices.insert_rows(0, vk);
+
+      arma::mat atom = (function.MatrixA()).col(k);
+      currentAtoms.insert_cols(0, atom);
+    }
+  }
+
+  /**
+   * Recover the solution coordinate from the coefficients of current atoms.
+   *
+   * @param n dimension of original solution space.
+   * @param x coefficients of current atoms.
+   */
+  arma::vec RecoverVector(const size_t n, const arma::vec& x)
```

Comment:
The compiler might be smart to optimize the copy away, but I was wondering if we should pass ``y`` as a function parameter?